runtime/io.h

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Defines API between compiled code and I/O runtime library.
+
+#ifndef FORTRAN_RUNTIME_IO_H_
+#define FORTRAN_RUNTIME_IO_H_
+
+#include "entry-names.h"
+#include <cinttypes>
+#include <cstddef>
+
+namespace Fortran::runtime {
+class Descriptor;
+class NamelistGroup;
+};
+
+namespace Fortran::runtime::IO {
+
+class IOStatementState;
+using Cookie = IOStatementState *;
+using ExternalUnit = int;
+using AsynchronousId = int;
+static constexpr ExternalUnit DefaultUnit{-1};  // READ(*), WRITE(*), PRINT
+
+extern "C" {
+
+#define IONAME(name) RTNAME(IO_##name)
+
+// These functions initiate data transfer statements (READ, WRITE, PRINT).
+// Example: PRINT *, 666 is implemented as the series of calls:
+//   Cookie cookie{BeginExternalListOutput(DefaultUnit)};
+//   OutputInteger64(cookie, 666);
+//   EndIOStatement(cookie);
+
+// Internal I/O initiation
+// Internal I/O can loan the runtime library an optional block of memory
+// in which the library can maintain state across the calls that implement
+// the internal transfer; use of these blocks can reduce the need for dynamic
+// memory allocation &/or thread-local storage.  The block must be sufficiently
+// aligned to hold a pointer.
+constexpr std::size_t RecommendedInternalIOScratchAreaBytes(
+    int maxFormatParenthesesNestingDepth) {
+  return 32 + 8 * maxFormatParenthesesNestingDepth;
+}
+
+// Internal I/O to/from a default-kind character scalar can avoid a
+// descriptor.
+Cookie IONAME(BeginInternalListOutput)(char *internal, std::size_t bytes,
+    void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+Cookie IONAME(BeginInternalListInput)(char *internal, std::size_t bytes,
+    void **scratchArea = nullptr, std::size_t scratchBytes = 0);
+Cookie IONAME(BeginInternalFormattedOutput)(char *internal, std::size_t bytes,
```

Comment:
The term formatted here is a little problematic.  The standard considers list directed and namelist I/O to be formatted I/O as much as I/O with an explicit format string.  Unfortunately, I don't have a good alternate name suggestion.  ExplicitFormat is clunky.  If someone can think of a nice term here, great.  Otherwise, I suppose this is ok.

Reply:
Agree.  Maybe "Edited" would be most in line with the terminology of the standard, but it would be less clear to most readers, I think.